iiif_prezi3/helpers/add_thumbnail.py

```diff
@@ -21,5 +22,66 @@ def add_thumbnail(self, image_url, **kwargs):
         self.thumbnail.append(new_thumbnail)
         return new_thumbnail
 
+    def create_thumbnail_from_iiif(self, url, preferred_width=500, **kwargs):
+        """Adds an image thumbnail to a manifest or canvas based on a IIIF service.
+
+        Args:
+            url (str): An HTTP URL which points at a IIIF Image response.
+            preferred_width (int, optional): the preferred width of the thumbnail.
+            **kwargs (): see ResourceItem.
+
+        Returns:
+            new_thumbnail (ResourceItem): The updated list of thumbnails, including the newly-created one.
+        """
+        image_response = ResourceItem(id=url, type='Image')
+        image_info = image_response.set_hwd_from_iiif(url)
+        context = image_info.get('@context', '')
+
+        if 'sizes' in image_info:
+            best_fit_size = min(
+                image_info['sizes'], key=lambda size: abs(size["width"] - preferred_width)
+            )
+            thumbnail_id = f"{url.replace('/info.json', '')}/full/{best_fit_size['width']},{best_fit_size['height']}/0/default.jpg"
+        else:
+            thumbnail_id = f"{url.replace('/info.json', '')}/full/full/0/default.jpg" if context == "http://iiif.io/api/image/2/context.json" else f"{url.replace('/info.json', '')}/full/max/0/default.jpg"
```

Comment:
Should this instead generate a image closer to the size provided? Assuming its not a level0 image.

Reply:
@glenrobson, I still need to refactor for this.  Is it safe to say that what's being suggested is if no sizes property exists and the image server is not level 0, this should return a thumbnail request for the exact width.  Is this right?

Reply:
if level0 then return max or full
else
  use exact side requested (for v2 width/ and for v3 width/(height)) height optional for v3 level 1 and above.